### ISSUE TITLE
Request for more intellectual free space count with btrfs

### ISSUE THREAD

**Issue by cryptofuture** [firehol/netdata#3095]
Basically low free space notification is useless now in case you run btrfs. Had downtime today due to no free space on `/` today.
You can read here: http://marc.merlins.org/perso/btrfs/post_2014-05-04_Fixing-Btrfs-Filesystem-Full-Problems.html
For example  `df` and `btrfs fi df` reports different space. I think metadata size should be in the count too. Also this means that all current disk stats in netdata (since notification based on [them](https://github.com/firehol/netdata/blob/master/conf.d/health.d/disks.conf)) are incorrect with Btrfs.

**Comment by ktsaou** [comment]
Yes you are right.

I tried a couple of times to find btrfs exposed metrics under `/proc` or `/sys`, without luck. So I guess the only solution to this problem is to run the `btrfs` command repeatedly to collect them.

Can you help us identify the metrics we should collect?

**Comment by cryptofuture** [comment]
`btrfs fi df /` (don't need root) output:
```
Data, single: total=7.31GiB, used=4.54GiB
System, DUP: total=8.00MiB, used=16.00KiB
Metadata, DUP: total=558.88MiB, used=177.11MiB
```
`btrfs fi show /` (needs root)
```
Label: none  uuid: axxx...
	Total devices 1 FS bytes used 4.73GiB
	devid    1 size 8.42GiB used 8.42GiB path /dev/vda2

```
As you can see used space is `Data` used + `System` used + `Metadata` used 
There is also which I miss `GlobalReserve`, but its accounted within Metadata
> GlobalReserve is an artificial and internal emergency space. It is used eg. when the filesystem is full. Its total size is dynamic based on the filesystem size, usually not larger than 512MiB, used may fluctuate.
> The global block reserve is accounted within Metadata. In case the filesystem metadata are exhausted, GlobalReserve/total + Metadata/used = Metadata/total.

`btrfs filesystem usage /` (show less information as not root)
without root:
```
Overall:
    Device size:		   8.42GiB
    Device allocated:		   8.42GiB
    Device unallocated:		   1.00MiB
    Device missing:		   8.42GiB
    Used:			   4.96GiB
    Free (estimated):		   2.69GiB	(min: 2.69GiB)
    Data ratio:			      1.00
    Metadata ratio:		      2.00
    Global reserve:		  64.00MiB	(used: 0.00B)
```
with root 
```
Overall:
    Device size:		   8.42GiB
    Device allocated:		   8.42GiB
    Device unallocated:		   1.00MiB
    Device missing:		     0.00B
    Used:			   4.96GiB
    Free (estimated):		   2.70GiB	(min: 2.70GiB)
    Data ratio:			      1.00
    Metadata ratio:		      2.00
    Global reserve:		  64.00MiB	(used: 0.00B)

Data,single: Size:7.31GiB, Used:4.61GiB
   /dev/vda2	   7.31GiB

Metadata,DUP: Size:558.88MiB, Used:177.28MiB
   /dev/vda2	   1.09GiB

System,DUP: Size:8.00MiB, Used:16.00KiB
   /dev/vda2	  16.00MiB

Unallocated:
   /dev/vda2	   1.00MiB
```
There are some information in `/sys/fs/btrfs/` but its mostly useless to the end user as disk geometry or sectorsize ```allocation  clone_alignment  devices  features  label  nodesize  sectorsize ```

### Inodes
Due dymanic inodes allocation there no interest in inodes at all.

So as far as I can say beside actual space real bummer to monitor is Metadata usage %. Not % but usage depending from metadata size itself.
Yes now df -h shows `/dev/vda2       8.5G  5.2G  2.6G  67% /` which is not bad since its higher than actual used space, but this is because I activate lzo compression, to free space up. Without compression stats are just opposite, and its easy to miss and run out of space.
Yes also might be interesting to show commit interval value for all fs somewhere in disks.

**Comment by Ferroin** [comment]
You can get the stats direct from `/sys/fs/btrfs`.  Each mounted filesystem has a directory there matching it's UUID, the free-space related stuff is in the `allocation` sub directory, and in fact provides much more info than the output of `btrfs fi df` (more in line with `btrfs fi usage`).  Within that directory, you have ond directory each for data, metadata, and system chunks with the following files being practically relevant:

* `total_bytes`: This is the equivalent of the 'total=' value  from `btrfs filesystem df`.  It shows how much space is allocated from the block-level allocator perspective.
* `bytes_used`: This is equivalent to the `used=` value from `btrfs filesystem df`.  It shows how much of the space reported by `total_bytes` is actually in use.
* `disk_total`: Like `total_bytes`, but accounting actual on-disk space (so if you've got things running in raid` mode, this will be double `total_bytes`).
* `disk_used`: Like `bytes_used`, but accounting actual on-disk space.

I'm not 100% certain about the meaning of all the other files (I think `bytes_pinned` has to do with space that's lost due to extent handling in reflinks, and `bytes_readonly` is usage that's only part of read-only snapshots, but I'm not certain , and have no idea about `bytes_may_use` or `bytes_reserved`).  You will also still need info from the output of `df`  (or whatever equivalent in `/proc` you use) to get info about the total size of the filesystem.

In addition, there is a subdirectory for each profile that is in use for that chunk type, with similar `total_bytes` and `bytes_used` directories.

The global reserve is tracked in total size only in the `global_rsv_size` file at the top level of the allocation directory, and I believe the free space within the reserve is tracked via `global_rsv_reserved`.

As far as inodes, @cryptofuture is correct, they're dynamically allocated, and there's no practical point in tracking them beyond figuring out how many files and directories are present (but they end up being per-subvolume, so even then it's not very useful).

It might be useful to get the free space estimate out of `btrfs filesystem usage`, (it's kind of complicated to calculate, and you can't do it with just the stuff from `/sys/fs/btrfs` because in-line compression impacts it).  

As far as what to alert on, just tracking metadata is not the way to go though, since metadata usage is dependent on workload and may not reflect the state of the filesystem.  The most general solution I can come up with is: to alert if the following equation approaches zero:

    total size reported by df - (data disk_used + metadata disk_used + system disk_used)

and the following is also approaching zero for at least one of data, metadata, and system:

    disk_used - disk_total

That combined condition will trigger an alert if you get close to exhausting space for a given type of chunk, which is what the primary -ENOSPC condition for BTRFS is (there are other possible conditions that can cause -ENOSPC, but that's the main one to worry about because if it happens for metadata, it's extremely difficult to recover).

**Comment by cryptofuture** [comment]
@Ferroin Thanks for more classification 
Found it exact location would be `/sys/fs/uuid/allocation/data` with
`bytes_may_use  bytes_pinned  bytes_readonly  bytes_reserved  bytes_used  disk_total  disk_used  flags  single  total_bytes  total_bytes_pinned`

**Comment by ktsaou** [comment]
This is excellent info! Thank you @Ferroin !

I have the following problem:

netdata reads `/proc/self/mountinfo` to find the mounted filesystems.

Now we have 2 options:

1. For each mounted btrfs filesystem, find the `/sys/fs/btrfs` info and add more charts about this btrfs subvolume. This means we have to correlate somehow the information available at `/proc/self/mountinfo` with the information at `/sys/fs/btrfs`.

or 

2. Don't correlate anything and just add a new section at the dashboard (like we do for NFS and ZFS), with the information found about BTRFS.

Given that the information at `/sys/fs/btrfs` is mainly disk space related, I think we should prefer option 1 above. If this is possible though, the only viable solution is option 2.

So, let me check how we do this correlation:

```sh
# cat /proc/self/mountinfo | grep btrfs
20 0 0:17 / / rw,relatime - btrfs /dev/sdd3 rw,compress=lzo,ssd,discard,space_cache,autodefrag,subvolid=5,subvol=/
49 20 0:17 /var/lib/docker/plugins /var/lib/docker/plugins rw,relatime - btrfs /dev/sdd3 rw,compress=lzo,ssd,discard,space_cache,autodefrag,subvolid=5,subvol=/var/lib/docker/plugins
50 20 0:17 /var/lib/docker/btrfs /var/lib/docker/btrfs rw,relatime - btrfs /dev/sdd3 rw,compress=lzo,ssd,discard,space_cache,autodefrag,subvolid=5,subvol=/var/lib/docker/btrfs
```

```sh
# tree /sys/fs/btrfs/
/sys/fs/btrfs/
├── 1faffa18-1db9-414e-8392-5643b5983fb7
│   ├── allocation
│   │   ├── data
│   │   │   ├── bytes_may_use
│   │   │   ├── bytes_pinned
│   │   │   ├── bytes_reserved
│   │   │   ├── bytes_used
│   │   │   ├── disk_total
│   │   │   ├── disk_used
│   │   │   ├── flags
│   │   │   ├── single
│   │   │   │   ├── total_bytes
│   │   │   │   └── used_bytes
│   │   │   ├── total_bytes
│   │   │   └── total_bytes_pinned
│   │   ├── global_rsv_reserved
│   │   ├── global_rsv_size
│   │   ├── metadata
│   │   │   ├── bytes_may_use
│   │   │   ├── bytes_pinned
│   │   │   ├── bytes_reserved
│   │   │   ├── bytes_used
│   │   │   ├── disk_total
│   │   │   ├── disk_used
│   │   │   ├── dup
│   │   │   │   ├── total_bytes
│   │   │   │   └── used_bytes
│   │   │   ├── flags
│   │   │   ├── total_bytes
│   │   │   └── total_bytes_pinned
│   │   └── system
│   │       ├── bytes_may_use
│   │       ├── bytes_pinned
│   │       ├── bytes_reserved
│   │       ├── bytes_used
│   │       ├── disk_total
│   │       ├── disk_used
│   │       ├── dup
│   │       │   ├── total_bytes
│   │       │   └── used_bytes
│   │       ├── flags
│   │       ├── total_bytes
│   │       └── total_bytes_pinned
│   ├── clone_alignment
│   ├── devices
│   │   └── sdd3 -> ../../../../devices/pci0000:00/0000:00:1c.3/0000:05:00.0/ata6/host5/target5:0:0/5:0:0:0/block/sdd/sdd3
│   ├── features
│   │   ├── big_metadata
│   │   ├── compress_lzo
│   │   ├── extended_iref
│   │   └── mixed_backref
│   ├── label
│   ├── nodesize
│   └── sectorsize
└── features
    ├── big_metadata
    ├── compress_lzo
    ├── default_subvol
    ├── extended_iref
    ├── mixed_backref
    ├── mixed_groups
    ├── no_holes
    ├── raid56
    └── skinny_metadata

12 directories, 52 files
```

The only correlation key I see is the disk (`/dev/sdd3`). But this means we will correlate it to all 3 mount points (btrfs root volume and 2 subvolumes).

So, I think the correlation is not possible and the only viable solution is option 2.
@Ferroin do you agree?

**Comment by Ferroin** [comment]
Keep in mind that BTRFS is both a filesystem _and_ a volume manager, just like ZFS.  I would say given that that the second approach is the best option.

That said, it is possible with some work to eliminate subvolumes from the list provided that they're mounted in-place.  Doing so requires some more inventive parsing of `/proc/self/mountinfo`, namely parsing parents of mounts to figure out which one is highest in the hierarchy (which should always be the highest point in the filesystem tree too, otherwise you get infinite recursion just like you can with self-referential directory symlinks).  Using the example you gave, the first line shows a mount ID of 20 (the first field), and the subvolumes both list that as their parent mount (the second field), so within that set the first line is the highest-level mount for that filesystem.

**Comment by ktsaou** [comment]
ok, here it is in PR #3150 

`/sys/fs/btrfs` provides all the values of:

```sh
# btrfs fi df /
Data, single: total=189.00GiB, used=180.34GiB
System, DUP: total=32.00MiB, used=48.00KiB
Metadata, DUP: total=6.50GiB, used=3.39GiB
GlobalReserve, single: total=512.00MiB, used=0.00B
```

The only value I didn't manage to find is `GlobalReserve.used`. On all my systems this is zero, so I guess the only way to find its source is to read the code.

Keep in mind `disk_total` and `total_bytes` are the same. I guess they would differ only on raid levels that provide redundancy (raid1, raid5, raid6). This means the `disk_*` versions of the metrics (total, used) are useless for our purpose.

IMPORTANT: **The problem is that /sys/fs/btrfs does not provide the capacity of the disks**.
My partition is 222GB:

```sh
# df -h /
Filesystem      Size  Used Avail Use% Mounted on
/dev/sdd3       222G  188G   29G  87% /
```

I couldn't find these 29GB in `/sys/fs/btrfs`.

As I understand it, btrfs auto-expands its footprint.

To test this, I created a simple btrfs raid0 on 4 loop devices (2GB each, 8 GB total):

```sh
btrfs fi df ./test-btrfs/
Data, RAID0: total=819.00MiB, used=256.00KiB
System, RAID1: total=8.00MiB, used=16.00KiB
Metadata, RAID1: total=409.56MiB, used=112.00KiB
GlobalReserve, single: total=16.00MiB, used=0.00B
```

It reports only 819MB, but:

```sh
# btrfs fi show test-btrfs 
Label: 'test-btrfs'  uuid: a2bc4480-5fbd-4246-b5ac-73cc9e18b194
	Total devices 4 FS bytes used 384.00KiB
	devid    1 size 2.00GiB used 614.31MiB path /dev/loop0
	devid    2 size 2.00GiB used 614.31MiB path /dev/loop1
	devid    3 size 2.00GiB used 212.75MiB path /dev/loop2
	devid    4 size 2.00GiB used 212.75MiB path /dev/loop3

# btrfs fi usage test-btrfs 
Overall:
    Device size:		   8.00GiB
    Device allocated:		   1.62GiB
    Device unallocated:		   6.38GiB
    Device missing:		     0.00B
    Used:			 512.00KiB
    Free (estimated):		   7.18GiB	(min: 3.99GiB)
    Data ratio:			      1.00
    Metadata ratio:		      2.00
    Global reserve:		  16.00MiB	(used: 0.00B)

Data,RAID0: Size:819.00MiB, Used:256.00KiB
   /dev/loop0	 204.75MiB
   /dev/loop1	 204.75MiB
   /dev/loop2	 204.75MiB
   /dev/loop3	 204.75MiB

Metadata,RAID1: Size:409.56MiB, Used:112.00KiB
   /dev/loop0	 409.56MiB
   /dev/loop1	 409.56MiB

System,RAID1: Size:8.00MiB, Used:16.00KiB
   /dev/loop2	   8.00MiB
   /dev/loop3	   8.00MiB

Unallocated:
   /dev/loop0	   1.40GiB
   /dev/loop1	   1.40GiB
   /dev/loop2	   1.79GiB
   /dev/loop3	   1.79GiB
```

Currently the charts show the following (the test btrfs with 4 loop disks):

![screenshot from 2017-12-18 00-34-56](https://user-images.githubusercontent.com/2662304/34084567-70a8550c-e38b-11e7-8cd3-e79ebcfefe46.png)

The stair step, are me writing a new 1GB file every a few seconds. The funny thing is that the `total` increases immediately, but the `used` increases after 30 seconds.

Here are my logical disk writes (every spike is 1GB written instantly by `dd`):

![image](https://user-images.githubusercontent.com/2662304/34084664-1a7ebd22-e38d-11e7-8ed3-4edd1cd9ed86.png)

Here is kernel dirty memory:

![image](https://user-images.githubusercontent.com/2662304/34084596-0f487ef8-e38c-11e7-9b79-98a7dd855414.png)

And these are the physical disk writes:

![image](https://user-images.githubusercontent.com/2662304/34084584-dcfd9050-e38b-11e7-9e77-d79496549df3.png)

If you look closely you will see that btrfs `total` increases when the `dirty` begins and btrfs `used` increases when the `dirty` clears and the physical write takes place.

This is the disk space are reported by the system:

![image](https://user-images.githubusercontent.com/2662304/34084626-843c95f0-e38c-11e7-9f3f-514dc132b26f.png)

and this is my `/tmp` (the place the loop devices are stored), which shows increased usage, as we add data to btrfs:

![image](https://user-images.githubusercontent.com/2662304/34084651-db586e9a-e38c-11e7-933d-2470a9aec3a5.png)

---

so I suggest to turn the btrfs charts to `stacked` charts and report `free` and `used` only (like the disk space charts, ignoring `disk_total` and `disk_used`). An alarm could be setup on these.

**But still the available disk space as an absolute number will be misleading!**

**Comment by ktsaou** [comment]
I forgot to mention: @Ferroin I would love to have your opinion on this. Do you agree with my suggestion? Any other/better solution?

**Comment by cryptofuture** [comment]
> The only value I didn't manage to find is GlobalReserve.used. On all my systems this is zero, so I guess the only way to find its source is to read the code.

You could find it in `man btrfs-filesystem`
> Global reserve is accounted within Metadata
> GlobalReserve is an artificial and internal emergency space. It is used eg. when the filesystem is full. 
> Its total size is dynamic based on the filesystem size, usually not larger than 512MiB, used may fluctuate.
>The global block reserve is accounted within Metadata. In case the filesystem metadata are exhausted, GlobalReserve/total + Metadata/used = Metadata/total.

> I suggest to turn the btrfs charts to stacked charts and report free and used only (like the disk space charts, ignoring disk_total and disk_used

I think its just what we need, due btrfs nature, btrfs  size could be increased anytime just as zfs pool.

**Comment by Ferroin** [comment]
@ktsaou My thoughts on all of this:

1. BTRFS does 'expand' over time in a way.  It uses a two-stage allocator, first allocating large regions of disk space for one type of block (data, metadata, or system), and then using a regular block allocator inside those regions.  However, the problematic situations arise when you run out of space at both the chunk level _and_ the block level.  Ideally, the percentage of the total device size used should be tracked too.  You can get the actual total space by looking at `/sys/fs/btrfs/<UUID>/devices/*/size` (the entries under the `devices` directory are symlinks to the sysfs entries for the devices themselves at the partition level, so the `size` attribute times the `queue/hw_sector_size` attribute will give you the device size in bytes).

2. The behavior you're seeing relating to the dirty data in RAM correlates with this and the fact that BTRFS is a COW filesystem.  Chunks get allocated (increasing the total) prior to anything getting written out from RAM, but the usage from the data being written out doesn't get accounted until the filesystem is fully updated and consistent again.  IOW, the process looks kind of like this when writing out a new file (greatly simplified and ignoring the more concrete layering approach taken in BTRFS):
    1. See if we need a new chunk, and if so allocate it.  <- This is where 'total' increase.
    2. Write out the blocks.
    3. Prepare the metadata for the blocks (actually happens in concert with step 2).
    4. Perform a copy-on-write update of the metadata at the next level up in the B-tree to reference the newly created metadata.
    5. Repeat step 4 until you get to the top of the B-tree.
    6. Update the superblock to point to the new B-tree root. <- This is where `used` increases.

2. I don't think that tracking the `disk_*` values on the same charts as `total_bytes` and `bytes_used` makes sense, they're inherently different measurements,  The `disk_*` values measure actual on-device usage, while the other two measure user-visible usage _after_ applying any profiles to things, and as a result, they will not match on most multi-device filesystems.  For the record, what traditional `df` reports is a pre-calculated value provided by the kernel that gives the theoretical size of the FS, and the actual on-disk usage based on that.  I would say the most sensible way to handle this is to track the `disk_*` values together with the sum total of device sizes on one chart with everything but the total device size stacked (so that you get a result visually similar to how filesystem usage is tracked by netdata), and then show `total_bytes` and `bytes_used` on one chart each for each chunk type (Data, Metadata, and System).  The first chart will then cover (at a glance) the most important aspect of usage (Is there a space shortage at the chunk level combined with one at the block level?), and the subsequent 3 provide some further breakdown of the info that's a mostly useful for interpreting the regular filesystem usage chart.

3. Free space in the sense that most people think of it as an absolute value is inherently problematic on BTRFS because:
    * You can have a mix of differing replication/striping/parity.  At the moment, you can only have an absolute combination of modes (roughly) equivalent to RAID1, RAID0, RAID10, RAID5 and RAID6, but the actual free space calculations even with just one mode used gets _really_ complicated in some cases (for example, if you're running an array that has disks of varying sizes), and once you factor in that different chunk types can be different modes, and you can have more than one mode for a given chunk type on the FS at a time, it gets _really_ crazy.
    * BTRFS supports transparent compression which can vary per-file (you can have filesystems where different parts use different algorithms), and can get variable results in terms of compression ratio (which makes it impossible to give an exact number for free space as most people understand it).
    *  Reference links add another type of 'sparse' file to account for.  If instead of the test you did creating 1 GB files with `dd`, you used `cp --reflink=auto` to create them beyond the first, you would only be using a little more than 1 GB of space on the FS, but would technically have a lot more space counted by the regular `du` program.
    * Small files get packed into the metadata blocks, so the location a file gets accounted can get very complicated to figure out.

    Given this, I don't think it makes sense to try and account 'free space' outside of the total disk usage sense that `df` does, unless you want to add some complicated logic to do the same calculations that `btrfs filesystem usage` does to give a free space estimate.

4. GlobalReserve is an odd case, it's technically accounted as well under the Metadata, but should be zero usage during normal operation.  It's a segment of the Metadata space reserved for COW operations when the FS is short on space at both the chunk and block levels so that you can either delete some stuff from the FS or reshape things at the device level (adding a new device, replacing an existing one with a larger one, etc) to get some more space.  I'm not entirely sure how `btrfs filesystem df` and `btrfs filesystem usage` are getting the info about it.

5. In terms of alerts, I'd revise what I stated earlier to just use the `disk_*` values and the sum total of device sizes as I listed above.

TL;DR:  I think most of it is sane, I would just change things so that you have all the `disk_*` values on one chart with a sum total of all the device sizes, and a separate chart each for Data, Metadata, and System showing just the `total_bytes` and `bytes_used` values, and keep in mind that knowing the total of all the device sizes is important for some of the more important alerts.

**Comment by ktsaou** [comment]
thanks!

So, I understand we should do these:

1. a stacked chart tracking *physical disk space allocation* with 3 dimensions:

   - **used** (`disk_used`)
   - **unallocated** (`(sum of disk sizes * hw_sector_size) - disk_total`)
   - **free** (`disk_total - disk_used`)

   An alarm should be added tracking **used** as a percentage of the total (sum of the 3 dimensions).

2. a stacked chart for **data** with dimensions **used** and **free** (total - used).

   An alarm should be added tracking **used** as a percentage of the total (sum of the 2 dimensions).

3. a stacked chart for **metadata** with dimensions **used**, **free** (total - used) and **reserved** (`GlobalReserve.size`).

   An alarm should be added tracking **used** as a percentage of the total (sum of the 3 dimensions).

4. a stacked chart for **system**, with dimensions **used** and **free** (total - used).

   An alarm should be added tracking **used** as a percentage of the total (sum of the 2 dimensions).

Right?

**Comment by ktsaou** [comment]
I think however that the alarms for data, metadata and system should only trigger if the first alarm (physical disk space %) is also triggered. I mean, there is no point to trigger and alarm on the allocated free space, if there is spare (unallocated) space available.

**Comment by Ferroin** [comment]
Yes, I think that is all correct, and yes, the alarms for Data, Metadata, and System (not that the System one is likely to ever trigger, System chunks are only used to hold the tree structure that tracks chunk allocations, so it's mostly empty on a majority of filesystems) should only trigger if the first one triggers (as it's the combination of the conditions that's an issue).

Also, if we can find some way to figure out actual GlobalReserve usage, then having an alarm that triggers if that is non-zero for an extended period of time would be nice too (that would help catch internal bugs in BTRFS relating to the allocator, but would also trigger when the other ones do too).

**Comment by ktsaou** [comment]
ok, done it.

Same test: 8GB in 4 disks, wrote files of 1GB each every a few seconds, until the disk if full, then deleted the files. The nice thing is that btrfs de-allocates the physical disk space once it is not used anymore (I didn't expect this - I thought it would require a rebalance or something, but it did it by itself seconds after the files were removed).

![screenshot from 2017-12-18 21-48-57](https://user-images.githubusercontent.com/2662304/34124922-57441b8e-e43d-11e7-8a45-723552a9a394.png)

Let me check the source code of btrfs for the source of `GlobalReserve.used`...

**Comment by ktsaou** [comment]
hm... the only problem now is that `hw_sector_size` is not available for partitions...

**Comment by Ferroin** [comment]
That's not _exactly_ what I had in mind, I was thinking more of having each of the `disk_*` values reported individually (the balance between the three can be useful information to figure out how the filesystem is used), but otherwise looks good.

Also, you've apparently got a relatively recent kernel running there, as BTRFS cleaned up the empty data chunks after you deleted the files (only more recent kernels do this).

As far as `hw_sector_size`, that gets a bit tricky.  The `dev` attribute will get you the major and minor for the partition, and the `partition` attribute will give you the partition number, which you can use to figure out what device to check in `/sys/dev/block`, but that only works reliably for cases where you're dealing with fewer than 15 partitions on the device...

**Comment by ktsaou** [comment]
Solved the partition thing (it was simple `/sys/fs/btrfs/UUID/devices/DEVICE/../queue/hw_sector_size`). So, netdata first tries it straight (without `/../`) and if that fails it tries with it.

btrfs-progs gets all the `fi df` values with an `ioctl`: https://github.com/kdave/btrfs-progs/blob/9fe889ac02b9c49b885c8999f5dd4e192697fa83/cmds-filesystem.c#L59-L102

> That's not exactly what I had in mind, I was thinking more of having each of the disk_* values reported individually (the balance between the three can be useful information to figure out how the filesystem is used), but otherwise looks good.

On the same chart? There will be quite a few of dimensions:

1. `unallocated`
2. `data used`
3. `data free`
4. `metadata used`
5. `metadata free`
6. `system used`
7. `system free`

I can easily do this, but what would be its use?

**Comment by ktsaou** [comment]
ok, it is not bad:

![image](https://user-images.githubusercontent.com/2662304/34128058-1395ecb8-e448-11e7-85ee-bc00a3c4fd30.png)

**Comment by DocMAX** [comment]
ja you also add support for multiple profiles (e.g. show raid5 and single usage)?

**Comment by vlvkobal** [comment]
@DocMAX could you elaborate? What information should we read and how to treat it? How it looks in the `/sys/fs/btrfs/*` files? What should we show on the charts?

**Comment by Ferroin** [comment]
@vlvkobal Based on the comment, it sounds like the request is for space usage information on a per-chunk-profile basis (BTRFS allows you to have different replication/striping/parity profiles for different chunk types, or even different chunks of the same type).

Relevant data is in `/sys/fs/btrfs/{UUID}/allocation/{chunk-type}/{profile-name}`. However, I’m really not sure what the best way to present this data is. The data itself is useful for a couple of reasons (right now mostly for monitoring profile conversions, but long-term possibly for any number of features that have been discussed on the mailing list), but actually presenting it is not trivial.

**Comment by DocMAX** [comment]
@Ferroin thank you for explaining. couldn't do it better.
yes, basically i had a profile conversion going on and wanted to monitor the process.
maybe there are more use cases.

**Comment by Ferroin** [comment]
Right now I think profile conversions are the only big one.

if the devs ever actually end up implementing per-subvolume profiles, this type of monitoring will become rather important.
